### DISCUSSION TITLE
How to identify network traffic from agent to server

### DISCUSSION THREAD

**Discussion by Rossco8** [gocd/gocd#11143] [General]
I would like to increase security on my setup of GoCD

Is there any way I can identify network traffic from the agent to the server? e.g. agent auto register key is always in a Header or parameter or some other unique property of the traffic (other than IP address)?

**Comment by chadlwilson** [comment]
Can you share how you hope to increase security by identifying the network traffic?

Auto-registration is optional so it cannot be in all traffic from the agent. To my knowledge, generally speaking auto-registration is used to allocate a UUID to the agent and issue the agent a token. The token is tied to a session. Subsequent calls from agent -> server supply the token in an `Authorization` header and are validated. There's also a cookie that is set whose purpose I don't personally yet fully understand :-)

If you just want to identify (authenticated+authorized) agent traffic in general you could monitor particular URI prefixes (`/go/remoting/*`) targetting the server. Or you could look for traffic with `X-Agent-GUID` header. Both of those are just the authenticated traffic - there are some routes which are open for the agent to bootstrap the latest agent code and register/authenticate itself.

**Comment by Rossco8** [comment]
**Can you share how you hope to increase security by identifying the network traffic?** 
This is my question - How to distinguish between traffic from an agent vs a malicious actor or my descktop accessing the server

auto-registration was just an example - any suggestions are obviously welcome

Can you tell me what to expect in `X-Agent-GUID` ? how to tell from a network rule if it is valid traffic? 

All I have right now is firewall rules allowing my desktop and agent IP addresses, I am seeking to understand if there is a better way

**Reply by chadlwilson** [reply]
I'm not sure why you would want another layer (a network or web application firewall) to validate things that are best validated at application layer, but I don't really understand your security model or what threats you might be trying to mitigate.

If you want to go down this path, perhaps you are best to packet capture from a known trusted agent, analyze HTTP traffic in Wireshark or similar and try to distill from that the expected behaviours. You can cross-reference from the route authorization rules [here](https://github.com/gocd/gocd/blob/master/server/src/main/java/com/thoughtworks/go/server/newsecurity/filterchains/AuthorizeFilterChain.java) if you like.

As I say [above](https://github.com/gocd/gocd/discussions/11143#discussioncomment-4542189), some legitimate traffic/routes from an agent are non-authenticated and open to all, others are authenticated, so if you try and block traffic from agents if it doesn't have `X-Agent-GUID` you will block bootstrapping+registration traffic. Routes in the above code with `ROLE_AGENT` will need the header, but there are a subset of `/admin` routes the agent uses which are non-authenticated.

When it is present, `X-Agent-GUID` should be a standard string serialization of a type 4 UUID, but this is not guaranteed to be the same format forever. Other than basic header format validation, only the server can really know if it's valid, by checking it's a known agent, and accompanied by a valid `Authorization` token issued and generated using a key only known to the server.

**Comment by Rossco8** [comment]
Thanks for the info, that answers the question. Given there are some unauthenticated routes to the server I will have to stick with allow-listing IP addresses. 

Can I raise an ER based on this? It would be great if there was a secret of some sort in the server's `cruise-config.xml` that I can manually add to each agent (similar to agent auto register key) that will always be present in agent-server traffic \ headers. This would greatly simplify locking down the server.

The GoCD project is awesome BTW, it is exactly what I needed!

**Reply by chadlwilson** [reply]
Thanks for the kind words. You can raise a request, but I'm not sure that means it will get done :)

At this stage I'm not quite sure what attack vector or threat model we are talking about here. Whitelisting traffic via IP address and/or using some special header are "solutions" of a sort, but I am not sure to which particular problem - and with which assumptions about deployment model. Is there an implicit assumption here that the GoCD server is somehow public on the internet (via a reverse proxy)? Are you concerned about access to the agent endpoints? What constraints are there in the environment?

If your goal is to limit traffic to agent-specific endpoints since these seem "higher risk", then generally that could be mitigated by
- having a different network path for agent -> server communication to "user web/api access" based on deploying your agents into a more highly trusted environment that sits somewhere close to the server (or at least not needing to access the server via the public internet). The server could be addressed by different domains/hostnames or subdomains or subroutes so the traffic can be distinguished and you can apply different rules to it.
- you could then use the distinct "for agents" route to make mutual TLS a requirement for the agent communication. There are some details [here](https://docs.gocd.org/current/installation/ssl_tls/end_to_end_transport_security.html) about configuring the agents with a client certificate that would be used for mTLS. You would need a PKI infrastructure to generate 1+ client certs  for your agents, and for your TLS-terminating reverse proxy (or Kubernetes ingress) to be configured to enforce and validate a client certificate for traffic.
- then on the distinct route intended for "user" access, you could use your reverse proxy to block HTTP paths intended only for agent communication and bootstrapping via this path.

Layering on mTLS would add the network-level protection you are looking for for agent requests. It's possible to use mTLS for the "user" access via a browser too, but can be a bit of a pain for folks to configure browsers for, so generally not so preferred by people in many-user scenarios. Most people/orgs will definitely not have their server public on the internet (this is generally not advisable despite https://build.gocd.org/go/pipelines being exactly that!), and instead use some other approach such as VPNs or other zero-trust network access solutions to provide access through some tunnelling approach.

If you can share a bit more about your deployment model, environment and constraints, and what hypothetical risks you are concerned about - I might be able to better understand options and possible trade-offs.

**Comment by Rossco8** [comment]
My environment is in Google Cloud. There are various GCP projects representing development, testing, production etc. each of these environments has an agent installed. 

There is a central GCP project hosting the GoCD server, source code and artifacts. There is no VPN, or any private network. The traffic between GCP projects is over public internet and APIs are all authenticated and protected with Google Cloud Armour. It is a very cloud native, public cloud design, hence the GoCD server IS visible over public internet.

At this stage GoCD Server is the only unauthenticated service. Given some predictable secret in the traffic I could easily protect the agent -> server traffic with a Cloud Armour rule. That just leaves the user -> server traffic wich I could happily live with IP allow-listing or add my own IAP protection in front of it

Thanks for the suggestions around dedicated routes and mTLS, I'll have a closer look at that.

In the world of pure cloud native design, it is always helpful to support authenticated traffic. In Googe Cloud, that might look like the agent including an Authorization header of the service account it is running as, which would be validated at the server with Google IAP. AWS has similar concepts

**Reply by chadlwilson** [reply]
Which login/authentication plugin are you using for GoCD, I wonder?

I don't think it's accurate to characterise GoCD Server traffic as unauthenticated. All web applications need some public endpoints/routes so they can load UI assets and allow the initial login/authn (even if they, say, redirect to an identity provider for login immediately) . All subsequent endpoints have traffic authenticated - just by the server itself - and these tend to vary in approach across tools (cookies, bearer tokens in headers, etc :-)

There's nothing to stop you putting something in front of GoCD that requires users to authenticate separately, and/or to forward-proxy the GoCD traffic from your agents to layer on a pre-shared secret which can be Google Cloud Armor validated if you want to go that path? Perhaps there's a way to use a sidecar proxy next to the GoCD agents to do that, perhaps even one which is GCP-aware and can stamp traffic with proof of service account? (I had seen such solutions in AWS land to, for example authenticate traffic to AWS ElasticSearch without the main code having to be "AWS IAM aware")

I guess I still don't understand what a predictable secret would look like and how that is a sufficient approach to solving your problem. I don't really see static pre-shared secrets/keys that could be validated in a "simple rule" as particularly "modern" cloud native design or great security practice in a general sense. Personally I don't really like the agent auto-registration key very much either.... but that's another story :-)

Speaking personally, the trouble with a lot of the "cloud native" solutions for identity and authentication of traffic like this are that there are too many options from different cloud providers, and expecting projects to integrate with them all and be aware is probably unrealistic except for projects that operate at serious scale. I'm not really sure what a "lowest common denominator" or sufficiently generic solution would look like (although am interested in exploring if there are simple patterns to follow).
